File: Python3/0705. Design HashSet.py
class MyHashSet:

    # def __init__(self):
    #     """
    #     Initialize your data structure here.
    #     """
    #     self.arr = [0]*(1<<20)


    # def add(self, key: int) -> None:
    #     self.arr[key] = 1

    # def remove(self, key: int) -> None:
    #     self.arr[key] = 0

    # def contains(self, key: int) -> bool:
    #     """
    #     Returns true if this set contains the specified element
    #     """
    #     return self.arr[key] == 1


    def __init__(self):
        """
        Initialize your data structure here.
        """
        self.arr = [[] for _ in range(857)]


    def add(self, key: int) -> None:
        if key not in self.arr[key%857]:
            self.arr[key%857].append(key) 

    def remove(self, key: int) -> None:
        if key in self.arr[key%857]:
            self.arr[key%857].pop(self.arr[key%857].index(key))

    def contains(self, key: int) -> bool:
        """
        Returns true if this set contains the specified element
        """
        return key in self.arr[key%857]


# Your MyHashSet object will be instantiated and called as such:
# obj = MyHashSet()
# obj.add(key)
# obj.remove(key)
# param_3 = obj.contains(key)

File: C++/0705. Design HashSet.cpp
class MyHashSet {
private:
    vector<list<int>> data;
    static const int base = 857;
    static int hash(int key){
        return key%base;
    }
public:
    /** Initialize your data structure here. */
    MyHashSet(): data(base) {}
    
    void add(int key) {
        int h = hash(key);
        for(auto i = data[h].begin();i!=data[h].end();i++){
            if((*i)==key) return;
        }
        data[h].push_back(key);
    }
    
    void remove(int key) {
        int h = hash(key);
        for(auto i = data[h].begin();i!=data[h].end();i++){
            if((*i)==key){
                data[h].erase(i);
                return;
            }
        }
    }
    
    /** Returns true if this set contains the specified element */
    bool contains(int key) {
        int h = hash(key);
        for(auto i = data[h].begin();i!=data[h].end();i++){
            if((*i)==key){
                return true;
            }
        }
        return false;
    }
};

/**
 * Your MyHashSet object will be instantiated and called as such:
 * MyHashSet* obj = new MyHashSet();
 * obj->add(key);
 * obj->remove(key);
 * bool param_3 = obj->contains(key);
 */
